### PR TITLE
[FlyDSL] Replace the split-K atomic combine with a workspace + reduce kernel

### PR DESCRIPTION
**Pull request by JohnQinAMD**
### What breaks

A FlyDSL split-K GEMM recorded into a CUDA graph can spin forever on replay. The device wedges — it does not return a wrong number and it does not recover. The same design also makes the op non-deterministic run to run, and costs accuracy, because partials are rounded to bf16 *before* they are accumulated.

### Root cause

It is a five-step chain. Each step is forced by the one above it.

**1. The splits are combined by a bf16 atomic add straight into C.** `kernels/splitk_hgemm.py:1059` on `main`:

```python
llvm.AtomicRMWOp(
    llvm.AtomicBinOp.fadd,        # dtype_ is bf16, packed vec2
    pair_ptr_v,                   # points into C
    pair_v,
    llvm.AtomicOrdering.monotonic,
    syncscope="agent",
)
```

**2. An atomic add needs an initialised target, so the kernel needs a barrier to produce one.** Somebody must zero the C tile before anyone adds to it, and everybody else must wait for that. That is the entire purpose of the semaphore/signal pair:

| step | code | why it exists |
| --- | --- | --- |
| first arriving block zeroes the C tile | `zero_c()` | gives the atomic somewhere to add into |
| it raises `signal[idx] = 1` | | announces the tile is initialised |
| every other block spin-waits on that signal | `split_k_barrier()` | adding before the zeroing lands would corrupt C |
| the last block to depart resets the counter to 0 | | restores the entry state for the next launch |

**3. That barrier only works if the counter is 0 when a launch begins.** Step 1 elects "first arrival" by `atomicAdd(...) == 0`, and step 4 elects "last departure" by comparing against a fixed arrival count. Both are false if the counter starts anywhere other than 0.

**4. The counter is zeroed exactly once — where it is allocated.** `gemm_kernels.py:700` on `main`:

```python
@functools.lru_cache(maxsize=128)                    # allocated once per (device, stream)
def _get_split_k_tensors(device, stream):
    semaphore = torch.zeros(...)                     # the only memset, ever
    signal = torch.zeros(...)
    return semaphore, signal
```

After that, the invariant is maintained solely by the kernel resetting the counter on its way out.

**5. Under CUDA graph capture, neither holds.** The `lru_cache` entry was allocated the first time the shape ran — during warmup, before capture began — so the recorded graph contains the kernel launch and **no memset**. Nothing re-establishes the entry state on replay. A counter left non-zero therefore stays non-zero forever: clearing it was never part of the graph.

For `small_m` that is fatal, because its reduction is gated on the same state it would have to clean (`kernels/small_m_hgemm.py:712-800`, condensed — the source is FlyDSL MLIR builder calls):

```python
712  prev = atomicAdd(semaphore[idx], 1)
725  first_arrival = (arrive_idx == 0)   # never true if the counter is dirty
728      zero_c_tile(...)                #   so the tile is never initialised
743      store(signal[idx], 1)           #   and the signal is never raised
753  while cur == 0:                     # every block waits here, forever
767      cur = load(signal[idx])
792  last_departure = (arrive_idx == 2 * SPLIT_K - 1)
799      semaphore[idx] = 0              # never reached, so the counter stays dirty
```

No first arrival → no signal → the wait never exits → the reset never runs. **The only path that could clean up is blocked by the state it would clean.**

The generic `hgemm` family survives a dirty counter because there its counter only elects who does the cleanup, with nothing conditioned on a zero start — it completes, with the counter drifting. That is why the currently dispatched family shows no symptom.

## What changed

The counter exists only to make in-place atomic accumulation safe. So rather than protect the counter, this removes the accumulation that needs it.

**In the main kernel: atomic add into C → plain store into a private slot.**

```python
# before — every split adds into the same C, after waiting for it to be zeroed
llvm.AtomicRMWOp(llvm.AtomicBinOp.fadd, pair_ptr_v, pair_v, ...)

# after — every split writes fp32 into a slot nobody else touches
ws_slot     = ks_idx * BLOCK_K_WARPS + wid_k
ws_row_base = ws_slot * m                       # workspace is [SLOTS, m, N]
row_valid   = m_global < m                      # masked: the reduce reads only real rows
WS_[(ws_row, ws_col)] = val                     # no atomics, no barrier, no zeroing
```

**New: `kernels/splitk_reduce.py` (145 lines) sums the slots.** One launch, no coordination:

```python
acc = WS_.vec_load((row, n_base), VEC)
for s in range_constexpr(1, SLOTS):             # fixed order -> bit-reproducible
    part = WS_.vec_load((row + m_rows * s, n_base), VEC)
    acc = arith.addf(acc, part)                 # summed in fp32, not bf16
if HAS_BIAS:
    acc = arith.addf(acc, extf(bias_vec))       # bias folded once, here
C_.vec_store((row, n_base), truncf(acc), VEC)   # C written exactly once
```

Ordering between the two launches comes from the stream — inside a graph, from the dependency edge between the two nodes. Nothing is shared between blocks, so there is no coordination state a replay can poison.

**On the host: a counter cache → a growable workspace.** `_get_split_k_tensors` and `_check_split_k_semaphore_capacity` are gone, replaced by `_get_split_k_workspace` sizing `slots * m * n` fp32 elements, plus `flydsl_splitk_prewarm_capture_workspace()` so the buffer can be sized *before* capture starts — growth inside a captured region is not allowed. This follows `aiter/tuned_gemm.py::_opus_prewarm_capture_workspace`, which does the same for opus.

**Deleted, not kept behind a flag:** `zero_c()`, the signal store, `split_k_barrier()`, the counter reset, the bf16 atomic accumulate, `SPLIT_K_SEMAPHORE_MAX_LEN`, and the semaphore/signal kernel ABI arguments. Both kernel files end up smaller than before (`splitk_hgemm.py` 1121 → 983, `small_m_hgemm.py` 1412 → 1242).

This is not a new pattern in this repo. `csrc/opus_gemm` already reduces exactly this way and says so in its ABI header — *"no atomic_add, no self-clear, no semaphore"*. That option is not open to the ASM a16w16 split-K GEMM, which ships as precompiled `.co` binaries and therefore has to fix its semaphore instead (ROCm/aiter#4494); the FlyDSL kernels are generated from the Python DSL, so the counter can simply go.

## Test plan

Three things need proving, and they use different tools.

### 1. Correctness, reproducibility and capture safety — `op_tests/test_flydsl_splitk_workspace.py`

New file. Seven `(shape, split_k, family)` cases, three assertions each:

| assertion | what it would catch |
| --- | --- |
| `test_workspace_matches_reference` | the reduce sums the wrong slots, or misses one |
| `test_workspace_is_deterministic` | any residual non-fixed-order accumulation |
| `test_graph_replay_matches_eager` | anything a captured replay carries between runs |

The third one deliberately corrupts the workspace between replays (`1e30`, `NaN`, `1e30`). Without that it would be worthless: a clean capture-and-replay passes on the atomic combine as well, so it has to seed the adversarial state that the old design could not recover from.

```bash
pytest -q op_tests/test_flydsl_splitk_workspace.py
```

### 2. Roofline sweep — same file, run directly

`@benchmark` sweep over 6 shapes x 3 split_k x 5 tilings, filtered to launchable configs. Its `TB/s` column counts the workspace round trip (`2 * slots * m * n * 4` bytes), so the design's real cost appears in the table instead of being hidden:

```bash
python3 op_tests/test_flydsl_splitk_workspace.py
```

### 3. Before/after A/B — two trees, one script

The legacy combine is deleted, so it cannot be selected with a flag. Instead the *same* script runs against two checked-out trees, mounted over the installed package:

```bash
# extract both trees
git archive origin/main         aiter/ops/flydsl | tar -x -C /tmp/base
git archive HEAD                aiter/ops/flydsl | tar -x -C /tmp/head

# run the identical sweep against each, one GPU, 30 timed iterations after 10 warmup
DP=/usr/local/lib/python3.12/dist-packages/aiter
for T in base head; do
  docker run --rm --ipc=host --shm-size=8g \
    --device=/dev/kfd --device=/dev/dri --group-add=video --group-add=render \
    -e HIP_VISIBLE_DEVICES=0 \
    -v /tmp/$T/aiter/ops/flydsl:$DP/ops/flydsl:ro -v $PWD:/work \
    --entrypoint bash "$IMAGE" -lc "python3 /work/ab_full.py /work/$T.json"
done
```

`ab_full.py` walks the same config space, times each config twice — once eager, once as a captured graph replayed 30 times — records max abs error against an fp32 torch reference, and writes JSON. Configs an unsupported tiling rejects are recorded as such and excluded from both sides, so the comparison is over the identical set. **71 configs measured on both trees.**

Full environment for all three:

```bash
IMAGE='vllm/vllm-openai-rocm:kimi-k3@sha256:5aa7e626ff73672f5ca7aae46754570488c23d33ca1ac90756a1d2d1a3fe099b'
docker run --rm --ipc=host --shm-size=16g \
  --device=/dev/kfd --device=/dev/dri --group-add=video --group-add=render \
  -e HIP_VISIBLE_DEVICES=0 --entrypoint bash "$IMAGE" -lc '
git clone -q https://github.com/ROCm/aiter.git /tmp/aiter && cd /tmp/aiter
git fetch -q origin pull/4622/head && git checkout -q --detach FETCH_HEAD
AITER_USE_SYSTEM_TRITON=1 GPU_ARCHS=gfx950 python3 setup.py develop
pytest -q op_tests/test_flydsl_splitk_workspace.py
python3 op_tests/test_flydsl_splitk_workspace.py
python3 aiter/ops/flydsl/test_flydsl_splitk_hgemm.py
'
```

## Test results

MI355X (gfx950). Ratios below are **this PR ÷ `main`, so < 1.0 is this PR being faster or more accurate.**

### Suites

| what | result |
| --- | --- |
| `pytest op_tests/test_flydsl_splitk_workspace.py` | **21 passed** (7 cases x 3 assertions) |
| perf sweep | **68 rows, `err` 0 on every one** |
| `test_flydsl_splitk_hgemm.py` (pre-existing) | **4/4 pass, `max_delta = 0.0000`** |

That last file's thresholds had been loosened to tolerate the atomic path — `max_delta_limit` of 32.0 and 8.0 with 1% of elements allowed to be arbitrarily wrong, i.e. 4 bf16 ULP. Tightened here to **one ULP at 100%** on all four cases, since the workspace path measures exactly 0.

### Accuracy — better on 71 of 71 configs, none worse

| | max abs error |
| --- | ---: |
| `main` (bf16 atomic accumulate) | 3.4760 |
| this PR (fp32 workspace) | **0.9994** |

0.9994 is half a bf16 ULP at that magnitude — correctly rounded, the floor for a bf16 output. The atomic path cannot reach it because it rounds each partial to bf16 *before* adding. Its error also varies run to run.

### Determinism

Bit-identical across two identical runs on **all 71 configs**. The atomic path differed in 27-75% of elements on 5 of 7 spot checks.

### Performance

| | median | mean | best | worst | faster on |
| --- | ---: | ---: | ---: | ---: | ---: |
| **graph replay** | **0.951x** | 0.911x | 0.664x | 1.179x | 45 / 71 |
| eager | 1.116x | 1.166x | 0.858x | 1.426x | 6 / 71 |

Split-K is selected for skinny-M decode shapes, and that is where the win lands:

| m | configs | graph median | slower |
| ---: | ---: | ---: | ---: |
| **1** | 39 | **0.805x** | 7 / 39 |
| 16 | 8 | 1.023x | 7 / 8 |
| 104 | 18 | 1.007x | 10 / 18 |
| 128 | 6 | 0.976x | 2 / 6 |

The remaining regressions are the second launch, not the kernel body. Timed in isolation inside a graph, the reduce node costs **1.0-1.3us regardless of workload** — 1.06us to move 640KB, 1.01us to move 57KB. It is launch-bound, and that floor matches the residual almost exactly: `m16_n7168_k512_spk2` regresses by 1.55us against a 1.27us reduce. Removing that cost means removing the second launch, which is the design.

Eager pays one extra host dispatch, a roughly constant 2-5us. That is implementation cost, not design cost.

### Capture safety

A clean capture-and-replay proves nothing here: it passes on the atomic combine too, because the counter only wedges the kernel once it has been left dirty. So both sides are measured with adversarial state seeded before capture. The object differs, necessarily — after this change there is no counter to seed, so the workspace is seeded instead:

| | `main` — counter seeded dirty | this PR — workspace filled with 1e30 / NaN |
| --- | --- | --- |
| clean capture, then replay | passes (non-discriminating) | passes |
| adversarial state, then replay | **hangs; child killed at a 120 s timeout** | **passes, bit-identical to eager** |

The `main` row is cited from earlier work on this defect, not produced by this PR's suite: it used a seeded-semaphore test against the unpatched tree, reproduced in the public image pinned above (`1 failed in 124.56s`, child `TimeoutExpired`). That test cannot have an "after" column here, because the semaphore it seeds no longer exists.

The `this PR` row is `test_graph_replay_matches_eager`, which replays four times and refills the workspace with `1e30`, `NaN`, `1e30` between them. Every replay stays bit-identical to eager, because the combine reads only slots it wrote in the same launch.

### A pre-existing bug this removes

Legacy split-K with `PERSISTENT_N_TILES > 1` in `small_m` does not merely lose precision, it produces wrong output. Reproduced on a pristine checkout of `main`, so it predates this change:

| code under test | split_k | max abs err | mismatch |
| --- | ---: | ---: | ---: |
| pristine `main`, run 1 | 4 | 31.99 | 1372 / 7168 |
| pristine `main`, run 2 | 4 | 40.04 | 3096 / 7168 |
| pristine `main` | 1 | 0.24 | **0** / 7168 |

Two runs of the same binary disagree, and `split_k=1` — which never enters the combine — is exact. `prepare_split_k_tile` / `split_k_barrier` are re-entered per tile against a counter shared with blocks that are at a different tile in their own sequence, so the "first arrival zeroes / last departure resets" invariants do not hold. Not fixed separately: the redesign removes the code that carries the bug.

## Overlap and limits

ROCm/aiter#4494 fixes the same class of defect on the ASM a16w16 split-K path, which cannot take this approach because it ships as precompiled binaries. The two are independent and touch different files.

`small_m` is not currently dispatched — `iter_small_m_registry_configs` is commented out at `gemm_kernels.py:24` and `aiter/configs/` holds 0 `small_m` entries. The hang this removes is therefore latent rather than something users hit today. The accuracy, determinism and capture-safety results apply to the dispatched `hgemm` family immediately.

## Tool assistance

Claude Opus 5 (1M context) assisted with the root-cause analysis, the redesign, the measurement harness and drafting this description.

### CODE REVIEW

File: op_tests/test_flydsl_splitk_workspace.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+            "tile_k": tile_k,
+            "split_k": self.split_k,
+            "block_m_warps": 1,
+            "block_n_warps": 4,
+            **self.overrides,
+        }
+
+    @property
+    def is_small_m(self) -> bool:
+        return self.overrides.get("kernel_family") == "small_m"
```
[the remark this answers is not in the document]
Comment (JohnQinAMD) [reply op_tests/test_flydsl_splitk_workspace.py]: Fixed in 7367529 — `MappingProxyType({})` instead of the `noqa`'d dict, so it's genuinely immutable rather than immutable-by-convention. Did the same to the shared `SMALL_M` dict just below, which had the same exposure across the four cases that use it. `|` on a mappingproxy returns a plain dict, so the per-case override sites are unchanged.

File: aiter/ops/flydsl/gemm_kernels.py [hunk 2]
```diff
@@ -36,7 +38,9 @@ def _get_dtypes():
     return dtypes
 
 
-SPLIT_K_SEMAPHORE_MAX_LEN = 256
+# Buffer offsets inside the FlyDSL kernels are i32 byte offsets, so keep the
+# fp32 workspace well below the 2GiB point where that arithmetic would wrap.
+SPLIT_K_WORKSPACE_MAX_BYTES = 1 << 31
```
[the remark this answers is not in the document]
Comment (JohnQinAMD) [reply aiter/ops/flydsl/gemm_kernels.py]: You're right that they disagreed, but the fix is the comment, not the limit — the comment had the mechanism wrong.

These kernels don't do raw pointer arithmetic for global access; they go through an AMD buffer descriptor. `num_records` is a 32-bit *byte* count (clamped to `0xFFFFFFFF` in `buffer_ops.create_buffer_resource*`) and the per-lane voffset is a 32-bit *element* offset scaled to bytes, so addressing wraps at 4GiB, not 2GiB. `1 << 31` is therefore already half the wrap point — deliberately conservative, not "exactly at the boundary". Lowering it would have cost headroom for no reason.

Rewrote the comment to state the actual bound, and fixed the same wrong claim in the `ValueError` message. 7367529.

File: aiter/ops/flydsl/gemm_kernels.py [hunk 3]
```diff
@@ -697,31 +689,120 @@ def _register_all_configs():
...
+
+    if torch.cuda.is_current_stream_capturing():
+        have = 0 if ws is None else ws.numel()
+        raise RuntimeError(
+            "FlyDSL split-K workspace must be sized before CUDA graph capture "
+            f"(need {elems} fp32 elements, have {have}). Run this shape eagerly "
+            "once, or call "
+            "aiter.ops.flydsl.gemm_kernels.flydsl_splitk_prewarm_capture_workspace(...) "
+            "on the capture stream, before capturing."
+        )
```
[the remark this answers is not in the document]
Comment (JohnQinAMD) [reply aiter/ops/flydsl/gemm_kernels.py]: Good catch, fixed in 7367529. The check and the allocation now run under the same `torch.cuda.device(device)` + `torch.cuda.stream(stream)` context, so they can't disagree. Note this is all behind the early return for an already-large-enough buffer, so it stays off the hot path.

`flydsl_splitk_prewarm_capture_workspace` had the same split, plus a related one you didn't flag: `torch.cuda.Stream()` binds to whichever device is *current*, so prewarming for a non-current device would have tested — and registered the workspace on — another device's stream. Same fix.

File: aiter/aot/flydsl/gemm.py [hunk 4]
```diff
@@ -295,8 +285,8 @@ def _compile_hgemm_to_cache(
         c_to_lds=c_to_lds,
         has_bias=has_bias,
     )
-    # FlyDSL JIT does not accept None for tensor slots; pass real buffers for
-    # optional bias and split-K sync tensors.
+    # FlyDSL JIT does not accept None for tensor slots; pass a real buffer for
+    # the optional bias.
     launch_bias = bias if has_bias else b
     _compile_executable_to_cache(
```
[the remark this answers is not in the document]
Comment (JohnQinAMD) [reply aiter/aot/flydsl/gemm.py]: Correct, and it defeated the point of AOT for every split-K config. Fixed in 7367529: `_compile_hgemm_to_cache` now compiles the reduce kernel too when `split_k > 1`.

The slot count is part of the layout contract between the main kernel, the reduce kernel and the precompiler, so rather than restate `split_k * (block_k_warps if hgemm else 1)` in a second file I moved it into `_split_k_workspace_slots` and had all three read it from there.

Verified on gfx950: `split_k=2` now emits two "compilation succeeded" lines (main + reduce), `split_k=1` still emits one.
